src/lib/netlist/plib/ppmf.h

```diff
@@ -168,31 +173,32 @@ namespace plib {
 		// extract the generic function and adjust the object pointer
 		void convert_to_generic(generic_function &func, mfp_generic_class *&object) const
 		{
+			// NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+			object = reinterpret_cast<mfp_generic_class *>(reinterpret_cast<std::uint8_t *>(object) + (m_this_delta >> 1));
 			if ((m_this_delta & 1) == 0)
 			{
-				// NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
-				object = reinterpret_cast<mfp_generic_class *>(reinterpret_cast<std::uint8_t *>(object) + (m_this_delta >> 1));
 				// NOLINTNEXTLINE(performance-no-int-to-ptr,cppcoreguidelines-pro-type-reinterpret-cast)
 				func = reinterpret_cast<generic_function>(m_function);
 			}
 			else
 			{
 				// NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
-				object = reinterpret_cast<mfp_generic_class *>(reinterpret_cast<std::uint8_t *>(object));
-
-				// otherwise, it is the byte index into the vtable where the actual function lives
-				// NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
-				std::uint8_t *vtable_base = *reinterpret_cast<std::uint8_t **>(object);
-				// NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
-				func = *reinterpret_cast<generic_function *>(vtable_base + m_function + m_this_delta - 1);
+				std::uint8_t *vtable_base = *reinterpret_cast<std::uint8_t **>(object) + m_function;
+				if (compile_info::arch::value == ci_arch::IA64)
+					// NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+					func = reinterpret_cast<generic_function>(uintptr_t(vtable_base));
+				else
+					// NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+					func = *reinterpret_cast<generic_function *>(vtable_base);
 			}
 		}
 
 		// actual state
-		uintptr_t               m_function;         // first item can be one of two things:
-													//    if even, it's a pointer to the function
-													//    if odd, it's the byte offset into the vtable
-		ptrdiff_t               m_this_delta;       // delta to apply to the 'this' pointer
+		uintptr_t m_function;   // first item can pointer to the function or a byte offset into the vtable
+		ptrdiff_t m_this_delta; // delta to apply to the 'this' pointer after right shifting by one bit
+		                        //    if even, m_function is a pointer to the function
+		                        //    if odd,  m_function is the byte offset into the vtable
+		                        //       or a byte offset into the function descriptors on IA64
```

Comment:
It’s always a byte offset into the vtable, it’s just the vtable may contain function descriptors rather than function pointers.

Reply:
I have separated out the raw processing code into `ppmf.cpp`. I have also taken first steps to clean the abi identification.

The code submitted now supports:

a) The unknown virtual inheritance case with specific code for MSVC for simple (i.e. scalar, void, reference, pointer) return types using the optimized code. This is the test case only supported under MSVC and which should return 7. Please have a look at the code. Basically m_vptr_offset is an index into a pointer table. It needs to be multiplied by `sizeof(void *)`. 

b) If you enable PPMF_EXPERIMENTAL complex return type member functions for MSVC will also use the optimized code path. This is explained in detail in the code in ppmf, including caveats and possible solutions.

c) Enabling PPMF_USE_MAME_DELEGATES will use `src/lib/util/delegate.h` as a replacement for `ppmf.h`. This allows to run the same tests (`nltool -c tests` from standalone build) as for the ppmf code.  

d) The code now also supports `clang-cl`